earth_enterprise/src/fusion/autoingest/sysman/InsetInfo.h

```diff
@@ -168,6 +168,60 @@ void CalcPacketGenInfo(const khTilespace            &tilespace,
                        const bool   is_overlay_terrain_proj,
                        const uint32 overlay_terrain_resources_min_level);
 
+// helper struct to make more generic calls from multiple locations possible
+// can be modified to allow for easier usage in possible new algorithms
+// for calculating overlap
+template <typename ProductAssetVersion>
+struct overlapEnvelope
+{
+    AssetDefs::Type& type;
+    khInsetCoverage gencov;
+    std::vector<const InsetInfo<ProductAssetVersion> *>& insets;
+    uint numInsets;
+    uint beginMinifyLevel;
+    uint endMinifyLevel;
+    uint level;
+    overlapEnvelope(AssetDefs::Type& _type,
+                    const khInsetCoverage& _gencov,
+                    std::vector<const InsetInfo<ProductAssetVersion> *>& _insets,
+                    uint _numInsets,
+                    uint _beginMinifyLevel, uint _endMinifyLevel, uint _level)
+
+   :                type(_type), gencov(_gencov), insets(_insets),
+                    numInsets(_numInsets),
+                    beginMinifyLevel(_beginMinifyLevel), endMinifyLevel(_endMinifyLevel),
+                    level(_level) {}
+};
+
+template <typename ProductAssetVersion>
+class OverlapCalculator
+{
+private:
+    std::vector<uint> neededIndexes;
```

Comment:
I think we can get away without storing neededIndexes permanently in this object for now since we're clearing it every time we use it. That will save us some memory, which is an important concern right now. Here's my recommendation.

In the `OverlapCalculator` class, remove the `neededIndexes` member variable.

Change `PreprocessForInset` from
```
    void PreprocessForInset()
    {
        neededIndexes.clear();
        CalculateOverlap();
    }
```
to
```
    std::vector<uint> PreprocessForInset()
    {
        std::vector<uint> neededIndexes;
        CalculateOverlap(neededIndexes);
        return neededIndexes;
    }
```

In BuildIndex, change:
```
    std::shared_ptr<OverlapCalculator<RasterProductAssetVersion>> calc =
        std::make_shared<OverlapCalculator<RasterProductAssetVersion>>(env);
    calc->PreprocessForInset();
    neededIndexes = std::move(calc->GetOverlapForLevel());
```
to
```
    std::shared_ptr<OverlapCalculator<RasterProductAssetVersion>> calc =
        std::make_shared<OverlapCalculator<RasterProductAssetVersion>>(env);
    neededIndexes = std::move(calc->PreprocessForInset());
```

There are a few other changes that will be necessary to make this work but I think those are the main ones. This will also allow you to get rid of the boolean parameter to GetOverlapForLevel and just always recalculate when that is called.